### PR TITLE
Add support for allocations

### PR DESCRIPTION
**Pull request by stbenjam**
Add support for allocations

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -50,6 +63,43 @@ resource "ironic_node_v1" "openshift-master-0" {
...
+  candidates = [
+    "${ironic_node_v1.openshift-master-0.id}",
+    "${ironic_node_v1.openshift-master-1.id}",
+    "${ironic_node_v1.openshift-master-2.id}",
+  ]
+
+  traits = [
+    "CUSTOM_FOO",
+  ]
+}
```
Comment (stbenjam) [inline README.md]: Example syntax for allocation is here

**Comment by stbenjam** [comment]
@dtantsur Could you have a look if you have a minute?

File: ironic/resource_ironic_allocation_v1.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	result, err := allocations.Create(client, allocationSchemaToCreateOpts(d)).Extract()
+	if err != nil {
+		return err
+	}
+
+	d.SetId(result.UUID)
+
+	// Wait for state to change from allocating
+	var state string
+	timeout := 1 * time.Minute
```
Comment (stbenjam) [inline ironic/resource_ironic_allocation_v1.go]: Is this a reasonable time period to wait for an allocation's state to change from `allocating` to either `error` or `active`?

Comment (dtantsur) [reply ironic/resource_ironic_allocation_v1.go]: More than. It's usually 1-3 seconds.

P.S.
I will check the whole patch tomorrow.

File: ironic/resource_ironic_allocation_v1.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	checkInterval := 5 * time.Second
+
+
+	for resourceAllocationV1Read(d, meta) ; state == "allocating" || state == ""; resourceAllocationV1Read(d, meta) {
```
Comment (dtantsur) [inline ironic/resource_ironic_allocation_v1.go]: nit: `state` cannot be empty in the API

File: ironic/resource_ironic_allocation_v1.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+	if err != nil {
+		return err
+	}
+
+	d.SetId(result.UUID)
+
+	// Wait for state to change from allocating
+	var state string
+	timeout := 1 * time.Minute
+	checkInterval := 5 * time.Second
```
Comment (dtantsur) [inline ironic/resource_ironic_allocation_v1.go]: 5 seconds is too much, most of allocations will be done in 1-2

File: ironic/resource_ironic_node_v1.go [hunk 5]
```diff
@@ -628,3 +664,14 @@ func (workflow *provisionStateWorkflow) changeProvisionState(target nodes.Target
 
 	return false, nodes.ChangeProvisionState(workflow.client, workflow.d.Id(), *opts).ExtractErr()
 }
+
+func changePowerState(client *gophercloud.ServiceClient, id string, target nodes.TargetPowerState) error {
+	if err := nodes.ChangePowerState(client, id, nodes.PowerStateOpts{
+		Target:  target,
+		Timeout: 30,
```
Comment (dtantsur) [inline ironic/resource_ironic_node_v1.go]: Are you sure you want to provide the timeout by default?

**Comment by stbenjam** [comment]
Sure, vendoring is pretty ugly.  That change is here: https://github.com/openshift-metalkube/terraform-provider-ironic/pull/14

I'm splitting out power state, as well.  This PR will have to go last, and tests will fail until both are merged.

**Comment by stbenjam** [comment]
Power state PR is here: https://github.com/openshift-metalkube/terraform-provider-ironic/pull/15

**Comment by stbenjam** [comment]
I'll rebase this one once the others are merged.

File: README.md [hunk 6]
```diff
@@ -50,6 +63,43 @@ resource "ironic_node_v1" "openshift-master-0" {
...
+  node_uuid   = "${ironic_node_v1.openshift-master-0.id}"
+  pxe_enabled = true
+  address     = "00:bb:4a:d0:5e:38"
+}
+```
+
+## Allocation
+
+The Allocation resource represents a request to find and allocate a Node
+for deployment. The microversion must be 1.52 or later.
```
Comment (dtantsur) [inline README.md]: We should really handle microversions ourselves..

File: ironic/provider.go [hunk 7]
```diff
@@ -20,13 +20,14 @@ func Provider() terraform.ResourceProvider {
 			"microversion": {
 				Type:        schema.TypeString,
 				Required:    true,
```
Comment (dtantsur) [inline ironic/provider.go]: This makes an API SIG member in me really, really sad.

File: ironic/resource_ironic_node_v1.go (hunk 5 above)
Comment (dtantsur) [reply ironic/resource_ironic_node_v1.go]: This is still a valid concern. Timeout requires a newer microversion and the server defaults may be good enough.

File: README.md (hunk 6 above)
Comment (stbenjam) [reply README.md]: To what extent? Should the provider only be valid for a minimum microversion?  Or do I really need to maintain knowledge in this code base of everything that a particular microversion supports?

File: ironic/provider.go (hunk 7 above)
Comment (stbenjam) [reply ironic/provider.go]: Could you elaborate? I’m happy to change it but I don't know what the proper thing to do is. This is the same approach as gopher cloud.

Comment (dtantsur) [reply ironic/provider.go]: The gopher cloud approach is not ideal, really.. I did not want to teach them how to manage their project, but microversions are not supposed to be exposed to users like that. It's largely our fault, as OpenStack API SIG, to fail to communicate it to people well enough. Please check my specifications on how we see it done: http://specs.openstack.org/openstack/api-sig/guidelines/sdk-exposing-microversions.html

File: README.md (hunk 6 above)
Comment (dtantsur) [reply README.md]: We don't need to know every microversion, but we can assume that if allocations are used, microversion must be at least 1.52. So if a user does not provide a microversion higher than that, use 1.52.

**Comment by stbenjam** [comment]
I filed https://github.com/openshift-metalkube/terraform-provider-ironic/issues/18, I'll open a separate PR to handle microversions as recommended for a high-level SDK.

**Comment by stbenjam** [comment]
Going to merge this, can iterate on next PR if anything needs changing.

**Pull request merged**